### ISSUE TITLE
Make hydra documentation respect access control

### ISSUE THREAD

**Issue by Benew** [api-platform/core#2719]
My api has public and private entries.

But the hydra and swagger documentations expose all of them.

As a result, a connected user that has access to only parts of the api will see the whole documentation which is not a good thing for security.

More importantly, the admin is broken in that case because it tries to fetch every resource in the hydra documentation.

**Comment by teohhanhui** [comment]
The documentation should be available even for operations that require authentication. There is nothing wrong with that. "Security by obscurity" is bad.

**Comment by teohhanhui** [comment]
> More importantly, the admin is broken in that case because it tries to fetch every resource in the hydra documentation.

Then that's a bug in the admin.

**Comment by Benew** [comment]
I agree. I first created an issue in the admin repository. I ended up overriding the documentation parser.
But still, I would like the ability to hide entrypoints from the swagger ui based on access control. It’s more of a feature request

**Comment by Benew** [comment]
Maybe it can be done in userland?

**Comment by er1z** [comment]
Just decorate documentation normalizer.

**Comment by dunglas** [comment]
It's a very frequent request. It would be nice to have it in core (with a flag to enable or disable this feature).

**Comment by dunglas** [comment]
Basically it means executing the security expression in the documentation normalizers.

**Comment by Benew** [comment]
I really need this and I’d need some directions to push a PR quickly. @dunglas what modifications and configuration options would you like to see?

**Comment by dunglas** [comment]
I suggest to inject the Security Expression executor in `DocumentationNormalizer`, and  to security expression of every resource class before [this line](https://github.com/api-platform/core/blob/master/src/Hydra/Serializer/DocumentationNormalizer.php#L79) and of every operation [here](https://github.com/api-platform/core/blob/master/src/Hydra/Serializer/DocumentationNormalizer.php#L219).

If the security expression can be executed (it doesn't depend of the current object but only use `is_granted()` for instance), then I would skip generating the documentation of not allowed operations.

The same logic must be adapted for the OpenAPI `DocumentationNormalizer`.

**Comment by gomcodoctor** [comment]
I am working on it,  i wanted to decorate DocumentationNormalizer but all functions are private !

**Comment by klobastov** [comment]
Is there any news on this topic?
